packages/expo-splash-screen/android/src/main/java/expo/modules/splashscreen/SplashScreenListenerActivity.kt

```diff
@@ -0,0 +1,5 @@
+package expo.modules.splashscreen
+
+interface SplashScreenListenerActivity {
+    fun onSplashScreenDismissed(): Unit
```

Comment:
I like this idea but I'm not sure I totally follow - the view provider classes don't inherit anything from the controllers (from what I can tell) - we could extend the SplashScreen.show to include a controller as an argument and implement `ManagedAppSplashScreenController.kt` which overrides the onDismissed method - is this sort of what you're getting at?

Reply:
I am sorry to say wrong and you are right, view providers does not inherit from controller.

still think on how we could not avoid introducing additional interface and here is yet another try by the `onViewDetachedFromWindow`.
when splash view is dismissed, the callback should be called.
here is the code snippet (taking `ManagedAppSplashScreenViewProvider.kt` as example)

```diff
--- a/android/expoview/src/main/java/host/exp/exponent/experience/splashscreen/ManagedAppSplashScreenViewProvider.kt
+++ b/android/expoview/src/main/java/host/exp/exponent/experience/splashscreen/ManagedAppSplashScreenViewProvider.kt
@@ -1,9 +1,14 @@
 package host.exp.exponent.experience.splashscreen

 import android.content.Context
+import android.graphics.Color
+import android.os.Handler
+import android.os.Looper
 import android.view.View
 import android.view.animation.AccelerateDecelerateInterpolator
 import android.view.animation.AlphaAnimation
+import android.widget.RelativeLayout
+import android.widget.TextView
 import com.squareup.picasso.Callback
 import com.squareup.picasso.Picasso
 import expo.modules.splashscreen.SplashScreenView
@@ -19,6 +24,8 @@ class ManagedAppSplashScreenViewProvider(
   private var config: ManagedAppSplashScreenConfiguration
 ) : SplashScreenViewProvider {
   private lateinit var splashScreenView: SplashScreenView
+  private val handler = Handler(Looper.getMainLooper())
+  private var timeoutRunnable: Runnable? = null

   companion object {
     private const val TAG: String = "ExperienceSplashScreenManifestBasedResourceProvider"
@@ -27,6 +34,7 @@ class ManagedAppSplashScreenViewProvider(
   override fun createSplashScreenView(context: Context): View {
     splashScreenView = SplashScreenView(context)
     configureSplashScreenView(context, config, null)
+    configureTimeoutHintView(context)
     return splashScreenView
   }

@@ -72,4 +80,34 @@ class ManagedAppSplashScreenViewProvider(
       }
     })
   }
+
+  private fun configureTimeoutHintView(context: Context) {
+    splashScreenView.addOnAttachStateChangeListener(object : View.OnAttachStateChangeListener {
+      override fun onViewAttachedToWindow(v: View?) {
+        if (timeoutRunnable == null) {
+          timeoutRunnable = Runnable { showTimeoutHintView(context) }
+          handler.postDelayed(timeoutRunnable!!, 2000)
+        }
+      }
+
+      override fun onViewDetachedFromWindow(v: View?) {
+        timeoutRunnable?.let { handler.removeCallbacks(it) }
+        timeoutRunnable = null
+      }
+    });
+  }
+
+  private fun showTimeoutHintView(context: Context) {
+    val textView = TextView(context)
+    textView.text = "Timeout!"
+    textView.setTextColor(Color.BLACK)
+
+    val layoutParams = RelativeLayout.LayoutParams(
+            RelativeLayout.LayoutParams.WRAP_CONTENT,
+            RelativeLayout.LayoutParams.WRAP_CONTENT)
+    layoutParams.addRule(RelativeLayout.CENTER_HORIZONTAL, RelativeLayout.TRUE)
+    layoutParams.addRule(RelativeLayout.ALIGN_PARENT_BOTTOM, RelativeLayout.TRUE)
+    layoutParams.bottomMargin = 100
+    splashScreenView.addView(textView, layoutParams)
+  }
 }
```

Reply:
no problem! I like this approach as well as its completely encapsulated by the view provider, but I don't think it would be the same for iOS, I think we'd have to provide a view controller as part of the view provider to get those view lifecycle events, which looks to be a good amount of refactoring.

another idea would be to have the viewprovider specify its splash screen controller via a `createSplashController` method, this way we could subclass the splash controller and get the behaviour we want. This would also be a bit of refactoring, but I think its the best option - what do you think? @Kudo

Reply:
I've updated with this implementation on iOS - let me know what you think!

Reply:
this approach is clever, opening the possibility to extend splash controller and cross-platformed.
thanks for the following up.